Change Catalog Dropdown design

elenatorro opened the pull request:
- Related issue: https://github.com/CartoDB/design/issues/1744

This PR includes:

- Use `Perfect Scrollbar` module (we've our [own fork](https://github.com/CartoDB/perfect-scrollbar) and we're using it in other views, like in Builder) to change the scrollbar style and make it compatible with different browsers. It also includes the feature that when hovering the scrollbar, it gets a bit wider.

- Set a max height for height-limited dropdowns in smaller screens using media queries, and apply the limit to both dropdowns in the catalog

- Adds a couple of small changes regarding responsive design (like setting a smaller font in the dropdown items, this needs to be reviewed)

**Screenshots:**

Tablet: 

![Screenshot 2019-11-12 at 10 40 31](https://user-images.githubusercontent.com/3824953/68661133-bae47d80-053a-11ea-9b35-9a2b0fe08939.png)

![Screenshot 2019-11-12 at 10 40 54](https://user-images.githubusercontent.com/3824953/68661135-bb7d1400-053a-11ea-9611-9803fdfa1980.png)

Mobile:

![Screenshot 2019-11-12 at 10 42 56](https://user-images.githubusercontent.com/3824953/68661136-bb7d1400-053a-11ea-8c9f-4c9f6cd1e096.png)
![Screenshot 2019-11-12 at 10 43 07](https://user-images.githubusercontent.com/3824953/68661137-bb7d1400-053a-11ea-8f81-86382fe41988.png)

Desktop (small):

<img width="1274" alt="Screenshot 2019-11-12 at 10 46 19" src="https://user-images.githubusercontent.com/3824953/68661138-bc15aa80-053a-11ea-8b36-f69b7669a0dd.png">

<img width="1276" alt="Screenshot 2019-11-12 at 10 46 30" src="https://user-images.githubusercontent.com/3824953/68661141-bc15aa80-053a-11ea-92c8-6f2c62811170.png">

alejandraarri commented on lib/assets/javascripts/new-dashboard/components/Catalog/CatalogDropdown.vue:
```diff
@@ -70,6 +74,10 @@ export default {
       type: Boolean,
       default: false
     },
+    searchDisabled: {
```
Would it make sense to turn round this property to `searchEnabled`? For me, is easier to understand: `searchEnabled: true/false` than `searchDisabled: false/true`.

elenatorro replied:
Right, I had the same doubt, but this makes more sense, I'll change it 👍

alejandraarri commented on lib/assets/javascripts/new-dashboard/styles/main.scss:
```diff
@@ -30,3 +30,6 @@
 @import 'hangar/main';
+
+//custom
+@import 'custom/perfect-scrollbar';
```
Instead of adding a new `custom` category, what about adding a `vendors` category for all external libraries related CSS?

elenatorro replied:
Sure! `vendors` is the correct name for these use cases 👍

alejandraarri commented on lib/assets/javascripts/new-dashboard/components/Catalog/CatalogDropdown.vue:
```diff
@@ -342,4 +354,20 @@ export default {
+  }
+}
+
+@media (max-height: $layout-tablet) {
```
Very clever!! I'm so used to use media queries for horizontal rules I hadn't think about this usage!

It would be nice to add another intermediate height media query to allow the dropdown render 5 items or so, if you feel like to. This is just an observation, because the dropdown renders 8 items or 3.

elenatorro commented:
PR updated! some notes here:

> Could we configure the scrollbar to be visible by default?

I tried this by setting `scrollToTop` to 1 to the dropdown list to make Perfect Scrollbar visible, but it does not work for that, I've been trying it also by overriding the styles but I couldn't :( 

> I'm not sure if I'd change the font size in the dropdowns for smaller devices

Reverted!

> Remove the Search on the Category dropdown

Understood and fixed! The only _issue_ here is that it does not detect window changes, which means, that if you increase or decrease the page height it will keep the initial state (showing or not the input), but if you reload the page it works. I think it doesn't worth it to listen to window changes 👍 

> Add another breakpoint:

Done! Now we've 3 breakpoints:

Small: shows up to 3 items
<img width="726" alt="Screenshot 2019-11-13 at 14 46 14" src="https://user-images.githubusercontent.com/3824953/68769869-5b16d100-0625-11ea-98c3-e43053f1b998.png">

Medium: shows up to 5 items

<img width="723" alt="Screenshot 2019-11-13 at 14 46 23" src="https://user-images.githubusercontent.com/3824953/68769933-77b30900-0625-11ea-9b73-14cb1f5e2123.png">

Default: shows up to 8 items

<img width="713" alt="Screenshot 2019-11-13 at 14 46 49" src="https://user-images.githubusercontent.com/3824953/68769925-7550af00-0625-11ea-8d58-3e6e34052b88.png">

> Click outside to close the dropdowns

I've encountered some errors when using the directive and I've removed it, not sure how to make it work better after trying several options.

elenatorro commented:
retest this please

alejandraarri commented on lib/assets/javascripts/new-dashboard/components/Catalog/CatalogDropdown.vue:
```diff
@@ -16,19 +16,16 @@
         <CatalogDropdownItem :option="searchFilter"/>
-        <button class="catalogDropdown__close" @click="reset">
+      </div>
+      <button class="catalogDropdown__close" @click="reset">
```
If you move the button outside the div it will be visible when the input is appearing, isn't it??

alejandraarri commented on lib/assets/javascripts/new-dashboard/components/Catalog/CatalogDropdown.vue:
```diff
@@ -93,6 +95,24 @@ export default {
     };
   },
   computed: {
+    maxItemsScroll () {
```
Wow! nice refactor!

elenatorro commented:
retest this please

alejandraarri commented on lib/assets/test/spec/new-dashboard/unit/specs/components/Catalog/CatalogDropdown.spec.js:
```diff
@@ -0,0 +1,163 @@
+import { shallowMount } from '@vue/test-utils';
```
🎉 Thanks for adding this, very detailed specs!!

alejandraarri commented on lib/assets/javascripts/new-dashboard/components/Catalog/CatalogDropdown.vue:
```diff
@@ -97,19 +129,53 @@ export default {
+    },
     showInput () {
-      return Object.keys(this.selected).length === 0;
+      return this.isOpen && Object.keys(this.selected).length === 0 && this.filteredOptions.length > this.maxItemsScroll;
```
Why not using `this.showScroll`?

elenatorro replied:
I had some doubts about it, it seems that by the time it's getting the `this.showScroll` computed value, it's `true`, while the comparison it's actually `false`, that's why I reverted it.

alejandraarri commented on lib/assets/javascripts/new-dashboard/components/Catalog/CatalogDropdown.vue:
```diff
@@ -182,20 +247,21 @@ export default {
+  --default-height-items: 8;
+  --medium-height-items: 5;
+  --small-height-items: 3;
+  --catalog-dropdown-item-height: 50px;
```
😍

alejandraarri replied:
Also, could you add, like 2px to `--catalog-dropdown-item-height` so the extra space in the dropdown to indicate scroll is more obvious? Right know, it can be seen as a UI error.

elenatorro replied:
I like the idea of setting the extra space more explicitly 👍

elenatorro commented:
retest this please

elenatorro commented:
Awesome 🎉 we're going to wait for design acceptance to merge it 👍 cc @csubira

elenatorro commented:
Hi @urbanphes 👋 could you please do the acceptance of this task? :) 

* Staging: https://catalog-dropdown.carto-staging.com/
* User: `catalog-dropdown`
* Pwd: `catalog-dropdown123`

urbanphes commented:
Nice job although I saw some things:

## 1. Extra Space
Between the title "Select a category" or "Select a country" and dropdown, there is an extra space

![image](https://user-images.githubusercontent.com/14546701/69138078-4ff8f080-0abe-11ea-9535-f27c87b53d98.png)

## 2. Icon missing
When a category or country is selected, the icon is missing

![image](https://user-images.githubusercontent.com/14546701/69139487-e9c19d00-0ac0-11ea-8b9a-dab4bbe4fe30.png)


## 3. Change hint 
On the country dropdown, the hint should say: "Select a country" and the color is grey.

![image](https://user-images.githubusercontent.com/14546701/69138361-cbf33880-0abe-11ea-86cd-51aa90ab9684.png)

## 4. Indicator scroll always visible
The indicator only appears when making scroll inside the dropdown, this one should appear always.

![image](https://user-images.githubusercontent.com/14546701/69138827-ae729e80-0abf-11ea-8164-205fdd55a2c9.png)
_Normal state_
![image](https://user-images.githubusercontent.com/14546701/69138849-b6324300-0abf-11ea-86a2-b48bed1a93b5.png)
_Scroll inside the dropdown_

## 5. Flags & Icons same size
On the Country dropdown, the Flags and the icons should the same size. Also, it will be awesome if we can implement all the flags.

![image](https://user-images.githubusercontent.com/14546701/69139224-5c7e4880-0ac0-11ea-9167-8b7bfce850f9.png)

## 6 You can't write anymore
In the country dropdown, it doesn't let put more characters

![image](https://user-images.githubusercontent.com/14546701/69139659-3efdae80-0ac1-11ea-81d9-e64ed990c555.png)


If you have doubts or need to clarify some point, tell me

csubira commented:
According to the acceptance below, the latest commits include:
1. Extra Space ✅ 
2. Icon missing ✅ 
3. Change hint ✅ 
4. Indicator scroll always visible 👎 -> Talked to @elenatorro and conclude that the way that the perfect scrollbar works it cannot be visible as desired.
5. Flags/icons same size ✅ 
6. You can't write anymore ✅ 

**Extra ball:** Add Description on Data / Catalo:  https://github.com/CartoDB/design/issues/1786

@urbanphes could you give a 👀 in staging:
Staging: https://carto-staging.com/
User: `user-catalog`
Pwd: `user-catalog123`
